runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+
+    /// Maximumm funding accepted, if this limit is reached, funding automatically is over.
+    pub max_amount: Balance,
+
+    /// Amount of stake required, possibly 0, to enter bounty as entrant.
+    pub entrant_stake: Balance,
+
+    /// Number of blocks from creation until funding is no longer possible. If not provided, then
+    /// funding is called perpetual, and it only ends when minimum amount is reached.
+    pub funding_period: Option<BlockNumber>,
+
+    /// Number of blocks from end of funding period until people can no longer submit
+    /// bounty submissions.
+    pub work_period: BlockNumber,
+
+    /// Number of block from end of work period until oracle can no longer decide winners.
+    pub judging_period: BlockNumber,
+
+    /// Funds provided by a bounty creator.
+    pub creator_funding: Balance,
+}
+
+/// Defines current bounty stage.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum BountyStage<BlockNumber> {
+    /// Bounty founding stage with starting block number.
+    Funding(BlockNumber),
+
+    /// A bounty was canceled.
+    Canceled,
+
+    /// A bounty was vetoed.
+    Vetoed,
+
+    /// A bounty funding was successful on the provided block.
+    /// The stage is set when the funding exceeded max funding amount.
+    MaxFundingReached(BlockNumber),
+}
+
+impl<BlockNumber: Default> Default for BountyStage<BlockNumber> {
+    fn default() -> Self {
+        BountyStage::Funding(Default::default())
+    }
+}
+
+/// Alias type for the Bounty.
+pub type Bounty<T> = BountyRecord<
+    BalanceOf<T>,
+    <T as frame_system::Trait>::BlockNumber,
+    <T as common::Trait>::MemberId,
+>;
+
+/// Crowdfunded bounty record.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct BountyRecord<Balance, BlockNumber, MemberId> {
+    /// Bounty creation parameters.
+    pub creation_params: BountyParameters<Balance, BlockNumber, MemberId>,
+
+    /// The current bounty stage.
+    pub stage: BountyStage<BlockNumber>,
+
+    /// The current unspent "cherry" balance.
+    pub cherry_pot: Balance,
+
+    /// The current unspent funding balance.
+    pub current_funding: Balance,
+}
+
+/// Balance alias for `balances` module.
+pub type BalanceOf<T> = <T as balances::Trait>::Balance;
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Bounty {
+        /// Bounty storage
+        pub Bounties get(fn bounties) : map hasher(blake2_128_concat) T::BountyId => Bounty<T>;
+
+        /// Double map for bounty funding. It stores member funding for bounties.
+        pub Funding get(fn funding_by_bounty_by_member): double_map hasher(blake2_128_concat)
+            T::BountyId, hasher(blake2_128_concat) MemberId<T> => BalanceOf<T>;
+
+        /// Count of all bounties that have been created.
+        pub BountyCount get(fn bounty_count): u32;
+    }
+}
+
+decl_event! {
```

Comment:
Can we make all of these have exhaustive parameters, so full extrinsic argument list.

Reply:
Done.